src/conduit/components/forms.cljs

```diff
@@ -0,0 +1,97 @@
+(ns conduit.components.forms
+  (:require [rum.core :as rum]
+            [clojure.string :refer [join]]
+            [citrus.core :as citrus]))
+
+(def initial-form-state {:data {} :errors {}})
+
+(def form-state (atom initial-form-state))
+
+(def email-regex #"[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?")
+
+(defn not-empty? [v]
+  (cond
+    (nil? v) false
+    (= "" v) false
+    :else true))
+
+(defn email? [v]
+  (not (nil? (re-matches email-regex (str v)))))
+
+(defn validate! [form-data form-errors validators]
```

Comment:
Validation function should be much simpler, there's no need for an internal state. Given a list of validators you can map over the list, execute each validator and return the result. Validation results should be set to a local state of the from component so it can display them.